File: kotlin/day14_part2_2023.kt
import java.io.File

data class Coord(val x: Int, val y: Int) {
    fun add(c: Coord) = Coord(x + c.x, y + c.y)
}

data class Grid(val width: Int, val height: Int, val data: MutableMap<Coord, Char>)

fun Coord.isInBounds(grid: Grid) = x in 0 until grid.width && y in 0 until grid.height

const val Empty = '.'
const val CubicRock = '#'
const val RoundRock = 'O'

val North = Coord(0, -1)
val West = Coord(-1, 0)
val South = Coord(0, 1)
val East = Coord(1, 0)

fun buildGrid(input: List<String>): Grid {
    val grid = Grid(input[0].length, input.size, mutableMapOf())

    input.forEachIndexed { y, line ->
        line.forEachIndexed { x, char ->
            if (char != Empty) {
                grid.data[Coord(x, y)] = char
            }
        }
    }

    return grid
}

fun Grid.toStringGrid(): String {
    var result = ""

    repeat(height) { y ->
        repeat(width) { x ->
            val coord = Coord(x, y)
            result += data[coord] ?: Empty
        }
        result += "\n"
    }

    return result
}

fun shiftSingleRock(grid: Grid, coord: Coord, dir: Coord) {
    if (grid.data[coord] == RoundRock) {
        var current = coord
        var before = coord.add(dir)

        while (!grid.data.containsKey(before) && before.isInBounds(grid)) {
            grid.data[before] = RoundRock
            grid.data.remove(current)

            current = before
            before = before.add(dir)
        }
    }
}

fun shiftRocks(grid: Grid, dir: Coord) {
    when (dir) {
        North, West -> {
            repeat(grid.width) { x ->
                repeat(grid.height) { y ->
                    shiftSingleRock(grid, Coord(x, y), dir)
                }
            }
        }
        South, East -> {
            for (x in grid.width - 1 downTo 0) {
                for (y in grid.height - 1 downTo 0) {
                    shiftSingleRock(grid, Coord(x, y), dir)
                }
            }
        }
    }
}

fun cycleRocks(grid: Grid) {
    shiftRocks(grid, North)
    shiftRocks(grid, West)
    shiftRocks(grid, South)
    shiftRocks(grid, East)
}

fun calculateGridKey(grid: Grid): Int {
    var key = 0

    repeat(grid.width) { x ->
        repeat(grid.height) { y ->
            val coord = Coord(x, y)
            if (grid.data[coord] == RoundRock) {
                key += coord.x + coord.y * grid.width
            }
        }
    }

    return key
}

fun calculateLoad(grid: Grid): Int {
    var load = 0

    repeat(grid.width) { x ->
        repeat(grid.height) { y ->
            val coord = Coord(x, y)
            if (grid.data[coord] == RoundRock) {
                load += grid.height - y
            }
        }
    }

    return load
}

fun solve(input: List<String>): Int {
    val numCycles = 1000000000
    val grid = buildGrid(input)
    val cache = mutableMapOf<Int, Int>()

    repeat(numCycles) { i ->
        val gridKey = calculateGridKey(grid)
        cache[gridKey]?.let { iStartCycle ->
            val remainingCycles = (numCycles - iStartCycle) % (i - iStartCycle)
            repeat(remainingCycles) {
                cycleRocks(grid)
            }
            return calculateLoad(grid)
        }
        cache[gridKey] = i

        cycleRocks(grid)
    }

    return calculateLoad(grid)
}

fun main() {
    val input = File("input.txt").readLines()
    println(solve(input))
}

File: julia/day14_part2_2023.jl

struct Coord
    x::Int
    y::Int
end

Base.:+(c1::Coord, c2::Coord) = Coord(c1.x + c2.x, c1.y + c2.y)

struct Grid
    width::Int
    height::Int
    data::Dict{Coord,Char}
end

isInBounds(coord::Coord, grid::Grid) = 0 <= coord.x < grid.width && 0 <= coord.y < grid.height

const Empty = '.'
const CubicRock = '#'
const RoundRock = 'O'

const North = Coord(0, -1)
const West = Coord(-1, 0)
const South = Coord(0, 1)
const East = Coord(1, 0)

function buildGrid(input::Vector{String})
    grid = Grid(length(input[1]), length(input), Dict{Coord,Char}())
    for (y, line) in enumerate(input)
        for (x, char) in enumerate(line)
            if char != Empty
                grid.data[Coord(x-1, y-1)] = char
            end
        end
    end
    return grid
end

function toString(grid::Grid)
    result = IOBuffer()
    for y in 0:grid.height-1
        for x in 0:grid.width-1
            coord = Coord(x, y)
            if haskey(grid.data, coord)
                print(result, grid.data[coord])
            else
                print(result, Empty)
            end
        end
        println(result)
    end
    return String(take!(result))
end

function shiftSingleRock!(grid::Grid, coord::Coord, dir::Coord)
    if get(grid.data, coord, Empty) == RoundRock
        current = coord
        before = current + dir
        while !haskey(grid.data, before) && isInBounds(before, grid)
            grid.data[before] = RoundRock
            delete!(grid.data, current)
            current = before
            before = before + dir
        end
    end
end

function shiftRocks!(grid::Grid, dir::Coord)
    if dir == North || dir == West
        for x in 0:grid.width-1
            for y in 0:grid.height-1
                shiftSingleRock!(grid, Coord(x, y), dir)
            end
        end
    else
        for x in grid.width-1:-1:0
            for y in grid.height-1:-1:0
                shiftSingleRock!(grid, Coord(x, y), dir)
            end
        end
    end
end

function cycleRocks!(grid::Grid)
    shiftRocks!(grid, North)
    shiftRocks!(grid, West)
    shiftRocks!(grid, South)
    shiftRocks!(grid, East)
end

function calculateGridKey(grid::Grid)
    key = 0
    for (coord, char) in grid.data
        if char == RoundRock
            key += coord.x + coord.y * grid.width
        end
    end
    return key
end

function calculateLoad(grid::Grid)
    load = 0
    for (coord, char) in grid.data
        if char == RoundRock
            load += grid.height - coord.y
        end
    end
    return load
end

function solve(input::Vector{String})
    numCycles = 1_000_000_000
    grid = buildGrid(input)
    cache = Dict{Int,Int}()
    for i in 0:numCycles-1
        gridKey = calculateGridKey(grid)
        if haskey(cache, gridKey)
            remainingCycles = (numCycles - i) % (i - cache[gridKey])
            for _ in 1:remainingCycles
                cycleRocks!(grid)
            end
            return calculateLoad(grid)
        end
        cache[gridKey] = i
        cycleRocks!(grid)
    end
    return calculateLoad(grid)
end

input = readlines("input.txt")
println(solve(input))
